File: src/robot.ts
interface Movement {
  [key: string]: () => void;
}
export class Robot {
  public x: number;
  public y: number;

  constructor(startX: number = 0, startY: number = 0) {
    this.x = startX;
    this.y = startY;
  }

  public process(commands: string[]): void {
    const movements: Movement = {
      N: () => this.moveNorth(),
      S: () => this.moveSouth(),
      E: () => this.moveEast(),
      W: () => this.moveWest(),
    };
    for (const command of commands) {
      movements[command.toUpperCase()]?.();
    }
  }

  public get position(): number[] {
    return [this.x, this.y];
  }

  private moveNorth(): void {
    if (this.y < 9) {
      this.y++;
    }
  }

  private moveSouth(): void {
    if (this.y > 0) {
      this.y--;
    }
  }

  private moveEast(): void {
    if (this.x < 9) {
      this.x++;
    }
  }

  private moveWest(): void {
    if (this.x > 0) {
      this.x--;
    }
  }
}

File: src/robot.rb
# typed: true
require 'sorbet-runtime'

class Robot
  extend T::Sig

  attr_accessor :x, :y

  sig { params(start_x: Integer, start_y: Integer).void }
  def initialize(start_x = 0, start_y = 0)
    @x = start_x
    @y = start_y
  end

  sig { params(commands: T::Array[String]).void }
  def process(commands)
    movements = {
      'N' => method(:move_north),
      'S' => method(:move_south),
      'E' => method(:move_east),
      'W' => method(:move_west)
    }

    commands.each do |command|
      movements[command.upcase]&.call
    end
  end

  sig { returns(T::Array[Integer]) }
  def position
    [@x, @y]
  end

  private

  sig { void }
  def move_north
    @y += 1 if @y < 9
  end

  sig { void }
  def move_south
    @y -= 1 if @y > 0
  end

  sig { void }
  def move_east
    @x += 1 if @x < 9
  end

  sig { void }
  def move_west
    @x -= 1 if @x > 0
  end
end
